lib/text.ml

```diff
@@ -100,26 +100,39 @@ let header_title ?(flavour = `Markdown) h =
 
 (* Toy change log parsing *)
 
-let change_log_last_entry ?flavour text =
+let rec change_log_last_entry ?flavour text =
   match head ?flavour text with
   | None -> None
   | Some (h, changes) -> (
       let title = header_title ?flavour h in
-      match String.take ~sat:Char.Ascii.is_graphic title with
-      | "" ->
-          Logs.app (fun m -> m "%S %S" h changes);
-          None
-      | version ->
-          let changes =
-            match
-              String.cuts ~sep:"\n" changes
-              |> List.map (String.drop ~rev:true ~sat:Char.Ascii.is_white)
-            with
-            | [] -> "(none)"
-            | "" :: lines -> String.concat ~sep:"\n" lines
-            | lines -> String.concat ~sep:"\n" lines
-          in
-          Some (Version.Changelog.of_string version, (h, changes)))
+      if String.(equal (Ascii.lowercase title) "changelog") then
+        (* Recurse for keepachangelog.com style *)
+        change_log_last_entry ?flavour changes
+      else
+        match String.take ~sat:Char.Ascii.is_graphic title with
+        | "" ->
+            Logs.app (fun m -> m "%S %S" h changes);
+            None
+        | version ->
+            let changes =
+              match
+                String.cuts ~sep:"\n" changes
+                |> List.map (String.drop ~rev:true ~sat:Char.Ascii.is_white)
+              with
+              | [] -> "(none)"
+              | "" :: lines -> String.concat ~sep:"\n" lines
+              | lines -> String.concat ~sep:"\n" lines
+            in
+            let version =
+              let fst = String.get version 0 in
+              let len = String.length version in
+              let last = String.(get version (len - 1)) in
+              if List.mem fst [ '('; '[' ] && List.mem last [ ')'; ']' ] then
+                (* Drop delimitors for keepachangelog.com style *)
+                String.with_index_range ~first:1 ~last:(len - 2) version
```

Comment:
I am not a big fan of index magic because it is tricky to read and a lot of code to begin with, only to realize that what it is doing is basically just trimming the delimiters.

Knowing that, I wondered how this could be rewritten in a more descriptive way since complaining without constructive critique is a bit silly and came up with this:

```ocaml
let version_delimiters = function '(' | ')' | '[' | ']' -> true | _ -> false in
let version = String.trim ~drop:version_delimiters version
```
This does the same thing but I think it is easier to read and understand. It also gets rid of the conditional because if there are no delimiters it is idempotent.

(One could argue that it accepts and removes the delimiters from `[0.0.1)` which is technically probably wrong but pragmatically this is probably good enough)

Reply:
Wasn't really happy with that part of the code myself. I've implemented your version.